src/stream/decoder.rs

```diff
@@ -121,16 +148,120 @@ impl<R: Read> Decoder<R> {
         }
 
         // And FILL IT!
-        let read = self.reader.read(&mut self.buffer)?;
+        let read = self.read_with_retry()?;
         unsafe {
             self.buffer.set_len(read);
         }
+
+        self.offset = 0;
         in_buffer.pos = 0;
         in_buffer.size = read;
-        // So we can't read anything: input is exhausted.
 
+        // So we can't read anything: input is exhausted.
         Ok(read > 0)
     }
+
+    fn read_with_retry(&mut self) -> Result<usize, io::Error> {
+        loop {
+            match self.reader.read(&mut self.buffer) {
+                Err(ref e) if e.kind() == io::ErrorKind::Interrupted => {},
+                otherwise => return otherwise
+            }
+        }
+    }
+}
+
+impl<R: Read> Decoder<R> {
+    /// This function handles buffer_refill state of the read operation
+    /// It returns true if read operation should be stopped and false otherwise
+    fn handle_refill(&mut self, hint: RefillBufferHint, 
+        in_buffer: &mut zstd_sys::ZSTD_inBuffer, out_buffer: &mut zstd_sys::ZSTD_outBuffer)-> Result<bool, io::Error> {
+
+        let refilled = match self.refill_buffer(in_buffer) {
+            Err(ref err) if out_buffer.pos > 0 && err.kind() == io::ErrorKind::WouldBlock => {
+                // underlying reader was blocked but we've already put some data into the output buffer
+                // we need to stop this read operation so data won' be lost
+                return Ok(true);
+            },
+            otherwise => otherwise
+        }?;
+
+        match hint {
+            RefillBufferHint::None => {
+                // can read again
+                self.state = DecoderState::Active
+            },
+            RefillBufferHint::FailIfEmpty => {
+                if refilled {
+                    // can read again
+                    self.state = DecoderState::Active;
+                }
+                else {
+                    // zstd keeps asking for more, but we're short on data!
+                    return Err(io::Error::new(io::ErrorKind::UnexpectedEof, "incomplete frame"));
+                }
+            },
+            RefillBufferHint::EndOfFrame => {
+                // at the end of frame
+                if refilled {
+                    // has more data - start new frame
+                    self.reinit()?;
+                    self.state = DecoderState::Active;
+                }
+                else {
+                    // no more data - we are done
+                    self.state = DecoderState::Completed;
+                    return Ok(true);
+                }
+            }
+        }
+
+        return Ok(false);
```

Comment:
Just `Ok(false)` -- `return` at the end isn't required and is not considered appropriate